src/bytes.rs

```diff
@@ -0,0 +1,294 @@
+        self.as_slice()
+    }
+}
+impl Deref for Bytes<'_> {
+    type Target = [u8];
+    fn deref(&self) -> &Self::Target {
+        self.as_slice()
+    }
+}
+impl Borrow<[u8]> for Bytes<'_> {
+    fn borrow(&self) -> &[u8] {
+        self.as_slice()
+    }
+}
+impl Debug for Bytes<'_> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("Bytes")
+            .field("data", &self.as_slice())
+            .field("owned", &!self.upgrade_will_allocate())
+            .field("shared", &!self.clone_will_allocate())
+            .finish()
+    }
+}
+impl Hash for Bytes<'_> {
+    fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
+        self.as_slice().hash(state)
+    }
+}
+impl<T: AsRef<[u8]>> PartialEq<T> for Bytes<'_> {
+    fn eq(&self, other: &T) -> bool {
+        self.as_slice() == other.as_ref()
+    }
+}
+impl Eq for Bytes<'_> {}
+impl<'a> From<&'a [u8]> for Bytes<'a> {
+    fn from(data: &'a [u8]) -> Self {
+        static VT: BytesVt = BytesVt {
+            release: None,
+            retain: Some(noop),
+        };
+        Bytes {
+            start: unsafe { &*data.as_ptr() },
+            len: data.len(),
+            data: data.as_ptr().cast(),
+            capacity: data.len(),
+            vtable: &VT,
+        }
+    }
+}
+#[cfg(feature = "alloc")]
+impl From<Arc<[u8]>> for Bytes<'static> {
+    fn from(data: Arc<[u8]>) -> Self {
+        extern "C" fn retain(this: *const (), capacity: usize) {
+            unsafe {
+                Arc::increment_strong_count(core::ptr::slice_from_raw_parts(
+                    this.cast::<u8>(),
+                    capacity,
+                ))
+            }
+        }
+        extern "C" fn release(this: *const (), capacity: usize) {
+            unsafe {
+                Arc::decrement_strong_count(core::ptr::slice_from_raw_parts(
+                    this.cast::<u8>(),
+                    capacity,
+                ))
+            }
+        }
+        static VT: BytesVt = BytesVt {
+            release: Some(release),
+            retain: Some(retain),
+        };
+        let capacity = data.len();
+        Bytes {
+            start: unsafe { &*data.as_ptr() },
+            len: data.len(),
+            data: Arc::into_raw(data) as *const (),
+            capacity,
+            vtable: &VT,
+        }
+    }
+}
+#[cfg(feature = "alloc")]
+impl<T: Sized + AsRef<[u8]> + Send + Sync> From<Arc<T>> for Bytes<'static> {
+    fn from(value: Arc<T>) -> Self {
+        extern "C" fn retain<T: Sized>(this: *const (), _: usize) {
+            unsafe { Arc::increment_strong_count(this.cast::<T>()) }
+        }
+        extern "C" fn release<T: Sized>(this: *const (), _: usize) {
+            unsafe { Arc::decrement_strong_count(this.cast::<T>()) }
+        }
+        let data: &[u8] = value.as_ref().as_ref();
+        Bytes {
+            start: unsafe { &*data.as_ptr() },
+            len: data.len(),
+            capacity: data.len(),
```

Comment:
Honestly, I plain dislike the `capacity` field, just had to introduce it to be conservative regarding `Arc::from_raw` on the most common `Arc<[u8]>`. It's really the only spot where it's currently useful, so I'd rather keep it semantically correct for the other types just in case we later decide to introduce Cow-mutability (something we ended up doing in Zenoh when shared memory became a concern for us)...